libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/OnlineRouteAlternativesSwitch.kt

```diff
@@ -0,0 +1,268 @@
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.ExperimentalMapboxNavigationAPI
+import com.mapbox.navigation.base.internal.accounts.UrlSkuTokenProvider
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin.Offboard
+import com.mapbox.navigation.base.route.RouterOrigin.Onboard
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.internal.accounts.MapboxNavigationAccounts
+import com.mapbox.navigation.core.internal.extensions.flowLocationMatcherResult
+import com.mapbox.navigation.core.internal.extensions.flowRouteProgress
+import com.mapbox.navigation.core.internal.extensions.flowRoutesUpdated
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.routeoptions.RouteOptionsUpdater
+import com.mapbox.navigation.core.trip.session.LocationMatcherResult
+import com.mapbox.navigation.utils.internal.logE
+import com.mapbox.navigation.utils.internal.logI
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withContext
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+
+private const val LOG_CATEGORY = "OnlineRouteAlternativesSwitch"
+
+/***
+ * This components replaces offline route by an online alternative when it's available. Under the
+ * hood it's requesting routes in cycle from the current position until the first successful result.
+ * Normally this job should be done by `RouteAlternativesObserver` that relies on platform
+ * reachability/connectivity notifications. However some environments
+ * doesn't have working reachability/connectivity API. This class is a temporary workaround for
+ * those environments until we reimplement route alternatives logic so that it handles
+ * offline-online switch even if platform's reachability API doesn't work.
+ *
+ * To enable automatic online-offline switch functionality, attach an instance of this class
+ * to [MapboxNavigation] using [MapboxNavigationApp.attach], followed by
```

Comment:
Wait, you just invoke `registerObserver`, you don't have to invoke attach/detach, right?

Reply:
I will replace by `registerObserver`